captum/_utils/sample_gradient.py

```diff
@@ -0,0 +1,176 @@
+from collections import defaultdict
+from enum import Enum
+from typing import Tuple, Union, cast, Iterable
+
+import torch
+from captum._utils.common import _format_tensor_into_tuples
+from torch import Tensor
+from torch.nn import Module
+
+
+def linear_param_grads(
+    module: Module, activation: Tensor, gradient_out: Tensor, reset: bool = False
+) -> None:
+    r"""
+    Computes parameter gradients per sample for nn.Linear module, given module
+    input activations and output gradients.
+
+    Gradients are accumulated in the sample_grad attribute of each parameter
+    (weight and bias). If reset = True, any current sample_grad values are reset,
+    otherwise computed gradients are accumulated and added to the existing
+    stored gradients.
+    """
+    if reset:
+        module.weight.sample_grad = 0  # type: ignore
+        if module.bias is not None:
+            module.bias.sample_grad = 0  # type: ignore
+    module.weight.sample_grad += torch.einsum(  # type: ignore
+        "n...i,n...j->nij", gradient_out, activation
+    )
+    if module.bias is not None:
+        module.bias.sample_grad += torch.einsum(  # type: ignore
+            "n...i->ni", gradient_out
+        )
```

Comment:
Does this assume that `gradient_out` and `activation` are 2 dimensional?  For a linear model they don't have to be 2 dimensional.

Reply:
This should work with any number of dimensions; the new test model also explicitly tests a 3D input to nn.Linear. The previous version only supported 2D inputs.

Reply:
@vivekmig, I asked because I tried n > 2 and it failed for linear model. I probably set something incorrectly.
```
import torch

a = torch.randn(4, 5)
b = torch.randn(4, 4)

# Passes
torch.einsum(  # type: ignore
        "n...i,n...j->nij", a, b
    ).shape

a = torch.randn(4, 5, 4)
b = torch.randn(4, 4, 4)

# Fails
torch.einsum(  # type: ignore
        "n...i,n...j->nij", a, b
    ).shape
```

Reply:
I think it works as long as all dimensions between the first and last are the same for both tensors. The '...' should serve to sum any dimensions in between the first and last. Your example seems to work fine when I run in colab.

In the previous version, it seems the tensors were these shapes:
```
a = torch.randn(4, 5, 4)
b = torch.randn(4, 4, 4)
```

I think that might have caused an issue, since the intermediate dimension should be the same for both, and the linear module only changes the size of the last dimension.

Reply:
<img width="1010" alt="Screen Shot 2022-02-21 at 12 30 22 PM" src="https://user-images.githubusercontent.com/12010601/155022648-8145f14e-43d3-496e-a0b6-42f12bfb384d.png">

That example is failing for me. Maybe it is dependent  on PT version ?

Reply:
Interesting, yes, from the documentation, seems like this might not be supported for versions before torch 1.8. I will double check and update the documentation, thanks for catching this!